### PR TITLE
fix(meta): drop `TableCatalog` in `cancel_stream_jobs`

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

An improvement on the behaviour of cancel stream job. We immediately drop the table catalog for a recovered stream job. If recovery is triggered, it will still cleanup the stream job.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/stream/stream_manager.rs [hunk 1]
```diff
@@ -657,18 +660,24 @@ impl GlobalStreamManager {
...
                     .await?;
-                if fragment.is_created() {
+                if table_fragments.is_created() {
                     Err(MetaError::invalid_parameter(format!(
                         "streaming job {} is already created",
                         id
                     )))?;
                 }
+                self.fragment_manager
+                    .drop_table_fragments_vec(&HashSet::from_iter(std::iter::once(
```
Comment (kwannoel) [inline src/meta/src/stream/stream_manager.rs:674]: This line is the only relevant part of the PR, the rest is refactoring.

**Comment by yezizp2012** [comment]
There's a corner case need to  taken into consideration: after the catalog is dropped, the cancel command could also be rejected if the cluster is right under recovery. If the clean operation in recovery has finished, the table fragments and related actors will be left in the system if there is no next recovery. We can solve it as what has been done for `drop`:
1. we should allow `cancel` command to be enqueued as `drop` when the queue is blocked, so that after recovery the `cancel` command could be processed.
https://github.com/risingwavelabs/risingwave/blob/c74817f797b5fba0eff25c9f4cfa95a817a0a392/src/meta/src/barrier/schedule.rs#L89-L100
2. pre apply the buffed `cancel` commands like `drop` during recovery, this's not necessary but useful.
https://github.com/risingwavelabs/risingwave/blob/c74817f797b5fba0eff25c9f4cfa95a817a0a392/src/meta/src/barrier/schedule.rs#L396-L415

**Review by yezizp2012 — APPROVED** [review]
LGTM! Btw have you tried any local tests for `cancel` during recovery?

**Comment by kwannoel** [comment]
> LGTM! Btw have you tried any local tests for `cancel` during recovery?

Hmm I'm not too sure how to test it. Any suggestions?

**Comment by yezizp2012** [comment]
> > LGTM! Btw have you tried any local tests for `cancel` during recovery?
> 
> 
> 
> Hmm I'm not too sure how to test it. Any suggestions?

For example:
1. create a materialized view in background mode.
2. kill one or some compute nodes before it gets completely.
3. cancel it to see if it can succeed.
4. restart the compute nodes to see whether the cluster is recovered and the materialized view is cancelled.

**Comment by kwannoel** [comment]
> > > LGTM! Btw have you tried any local tests for `cancel` during recovery?
> > 
> > 
> > Hmm I'm not too sure how to test it. Any suggestions?
> 
> For example:
> 
>     1. create a materialized view in background mode.
> 
>     2. kill one or some compute nodes before it gets completely.
> 
>     3. cancel it to see if it can succeed.
> 
>     4. restart the compute nodes to see whether the cluster is recovered and the materialized view is cancelled.

Added a deterministic test for it. The test will fail as expected on latest main:
```
2022-09-26T00:27:45.518443Z  INFO integration_tests::recovery::background_ddl: recreating mv
2022-09-26T00:27:45.522084Z  INFO pgwire_query_log: mode=(simple query) session=0 status=ok time=0ms sql=SET BACKGROUND_DDL=false
2022-09-26T00:27:45.543535Z  INFO pgwire_query_log: mode=(simple query) session=0 status=err time=9ms sql=CREATE MATERIALIZED VIEW mv1 as values(1)
Error: db error: ERROR: Failed to run the query

Caused by these errors (recent errors listed first):
  1: gRPC request to meta service failed: Internal error
  2: table is in creating procedure: 1003


Caused by:
    ERROR: Failed to run the query
    
    Caused by these errors (recent errors listed first):
      1: gRPC request to meta service failed: Internal error
      2: table is in creating procedure: 1003
```

With this PR, it will no longer fail this scenario.

**Comment by kwannoel** [comment]
Main cron test TLE is expected. Today's main-cron failed as well.

**Pull request merged**